### DISCUSSION TITLE
✔ CSP Limits Recommendation

### DISCUSSION THREAD

**Discussion by riramar** [oshp/oshp-tracking#29] [Q&A] [answered]
I'm looking for the recommendation in case of too many subdomains in the CSP response header.
For example, when it's recommended to use *.example.com since the example.com subdomains allowed are too many.
I double checked here https://www.w3.org/TR/CSP3/ but couldn't find anything about it.

**Comment by righettod** [comment]
Nice question indeed 👍 

If I have understood, your point is, for example regarding the following site with the following set of sub domains:

```text
sub1.example.com
sub2.sub1.example.com
sub3.sub2.sub1.example.com
```

How to use CSP to only allow content from `sub3.sub2.sub1.example.com` ?

**Reply by riramar** [reply]
> Nice question indeed 👍
> 
> If I have understood, your point is, for example regarding the following site with the following set of sub domains:
> 
> ```
> sub1.example.com
> sub2.sub1.example.com
> sub3.sub2.sub1.example.com
> ```
> 
> How to use CSP to only allow content from `sub3.sub2.sub1.example.com` ?

Not exactly. I meant in case of the CSP header is too big that cause problems in certain web frameworks like:

`Content-Security-Policy: default-src 'self'; img-src image1.example.com image2.example.com ...; media-src media1.example.com media2.example.com ...; script-src script1.example.com script2.example.com ...`

AFAIK the HTTP RFC do not specify a limit and some frameworks can breaks when receiving a big header.

**Comment by riramar** [comment]
BTW I posted the same question here https://lists.w3.org/Archives/Public/public-webappsec/2024Oct/0010.html to try get some info from the browser devs.

**Comment by righettod** [comment]
Thanks for the clarification, I will take a look as well and keep you posted with the results.

**Comment by righettod** [comment]
@riramar Do you have received any feedback about your question?

🤔If I'm not wrong, in case of a large CSP policy sent by the app server or the WAF or the reverse proxy, it is the browser or any network device handling the HTTP response that will *cut* or *alter* the CSP received.  At the framework level, it is just a string.

Do you have an example to allow me to better understand the issue, and its context, in order that I work on a proposal?

Thanks a lot in advance for your insights 😃

**Comment by riramar** [comment]
Hi @righettod 

I don't have a specific example for that. I found this paper from 2016 on Google which seems to point to the right direction.

https://research.google/pubs/csp-is-dead-long-live-csp-on-the-insecurity-of-whitelists-and-the-future-of-content-security-policy/
**We expect that that the combination of a nonce-based approach and the ’strict-dynamic’ keyword will allow developers and organizations to finally enjoy real security benefits offered by the Content Security Policy.**

The entire paper PDF can be found here: https://dl.acm.org/doi/pdf/10.1145/[2976749.2978363](https://dl.acm.org/doi/pdf/10.1145/2976749.2978363)

Regards,
Ricardo Iramar

**Comment by righettod** [comment]
👨‍💻I performed the following test.

📝Test page generating a large CSP based on a number of subdomains to add into the `img-src` directive:

```php
<?php
ini_set('display_errors', 1);
ini_set('display_startup_errors', 1);
error_reporting(E_ALL);
$csp="default-src 'self'; img-src 'self'";
$cnt=intval($_GET["c"]);
for ($x = 0; $x <= $cnt; $x++) {
	$csp .= " https://sub$x.righettod.eu";
}
$csp .= ";";
header("Content-type: text/html; charset=utf-8");
header("Content-Security-Policy: $csp", True, 200);
?>
<!DOCTYPE html>
<html>
<head>
	<title>test</title>
</head>
<body>
	<b>CSP policy generated (length: <?php echo(strlen($csp)); ?> characters): </b><br>
	<code><?php echo($csp); ?><br> 
	<script>alert("test xss");</script>
</body>
</html>
```

🌏Page exposed on Internet via a [segfault](https://shell.segfault.net/) instance:

![image](https://github.com/user-attachments/assets/2cfd062d-7530-41e6-ad4c-3e339206f343)

💻Test of generation of 1000 subdomains and loading of the page in the Firefox **132.0.2** (last release):

![image](https://github.com/user-attachments/assets/354ad5c2-41c5-43ec-8f67-5d81a01f41f5)

✅CSP correctly loaded and applied to block the inline JS code.

💻Same test in Chromium **131.0.6755.0**:

![image](https://github.com/user-attachments/assets/23b34e6c-2371-4182-b31a-8ccdd3884b3c)

✅CSP correctly loaded and applied to block the inline JS code.

💻Same tes in Edge **131.0.2903.51** (last release):

![image](https://github.com/user-attachments/assets/beb03440-e0e8-46ca-b47c-9daaecc946b3)

✅CSP correctly loaded and applied to block the inline JS code.

👨‍💻I tested to reach the limit.

💻For Edge and Chromium it was when CSP was above 261926 characters:

![image](https://github.com/user-attachments/assets/450761c8-4a2c-48f6-8255-ac7a401abd4d)

![image](https://github.com/user-attachments/assets/6a01ba1d-f131-4597-8ea3-7964b8922838)

💻For FF it was when CSP was above 392995 characters:

![image](https://github.com/user-attachments/assets/fc309a2a-97f4-47ec-a167-5265abc3686a)

👀Above the limit specified, the error was the following on the browser side (response correctly generated and sent by the php server):

![image](https://github.com/user-attachments/assets/4d2139a9-9502-455f-8a4a-9158124def20)

![image](https://github.com/user-attachments/assets/69ad2857-14eb-4a77-b04e-740f2788f7e6)

**Comment by righettod** [comment, marked as answer]
Based on the test performed, for me, modern browsers like Chromium based ones or FF supports sufficient size to specify a large CSP in case of need.

**Comment by riramar** [comment]
Agreed! I think in that case we can close this topic.
Thanks a lot for the tests.

**Reply by righettod** [reply]
Thanks to you for this very interesting question 💯
